src/jit/optimizer.cpp

```diff
@@ -1574,177 +1586,149 @@ void Compiler::optFindNaturalLoops()
-                        goto NO_LOOP;
+                        BasicBlock* pred = predIter->flBlock;
+
+                        unsigned int testNum = positionNum(pred);
+
+                        if ((testNum < top->bbNum) || (testNum > bottom->bbNum))
+                        {
+                            // Pred is out of loop range
+                            if ((pred == head) && (block == entry))
+                            {
+                                // This is the single entry we expect.
+                                continue;
+                            }
+
+                            // There are multiple entries to this loop, don't consider it.
+                            goto NO_LOOP;
+                        }
+
+                        if (pred == entry)
+                        {
+                            // We have indeed found a cycle in the flow graph.
+                            foundCycle = true;
+                            assert(loopBlocks.isMember(pred->bbNum));
+                        }
+                        else if (!loopBlocks.testAndInsert(pred->bbNum))
+                        {
+                            // Add this pred to the worklist
+                            worklist.push_back(pred);
+                        }
                     }
                 }
 
-                if (!canIterateLoop)
+                if (!foundCycle)
                 {
+                    // There isn't actually a loop between TOP and BOTTOM
                     goto NO_LOOP;
                 }
 
-                /* Double check - make sure that all loop blocks except ENTRY
-                 * have no predecessors outside the loop - this ensures only one loop entry and prevents
-                 * us from considering non-loops due to incorrectly assuming that we had a back edge
-                 *
-                 * OBSERVATION:
-                 *    Loops of the form "while (a || b)" will be treated as 2 nested loops (with the same header)
-                 */
-
-                for (loopBlock = top; loopBlock != bottom->bbNext; loopBlock = loopBlock->bbNext)
+                if (!loopBlocks.isMember(bottom->bbNum))
                 {
-                    if (loopBlock == entry)
-                    {
-                        continue;
-                    }
-
-                    for (predTop = loopBlock->bbPreds; predTop != nullptr; predTop = predTop->flNext)
-                    {
-                        if (predTop->flBlock->bbNum < top->bbNum || predTop->flBlock->bbNum > bottom->bbNum)
-                        {
-                            // noway_assert(!"Found loop with multiple entries");
-                            goto NO_LOOP;
-                        }
-                    }
+                    // The "back-edge" we identified isn't actually part of the flow cycle containing ENTRY
+                    goto NO_LOOP;
```

Comment:
You're following correctly.  Our loop detection algorithm (both before and after this change) already walks all inclusive blocks of each loop and is hence quadratic, regardless of this bail-out.  We cap the number of loops we track at 16, which I thought was what prevented this from running away; I've just now discovered that the code didn't actually bail out but instead kept looking for loops that it was doomed to discard, so I've added a bail-out in the update I just pushed.

Regarding "Do we need a similar check for TOP? Seems like we want to ensure that TOP and BOTTOM define the lexical extent of the loop so that the range-inclusive checks later work out" -- yes, we did, thanks.  I was thinking that finding BOTTOM in the set was sufficient because it has TOP as a successor, but I had that backwards because the walk is visiting predecessors -- so finding TOP is sufficient because it has BOTTOM as a predecessor, and I've pushed an update accordingly.